packages/indicators/src/IndicatorApi.ts

```diff
@@ -26,39 +29,91 @@ export class IndicatorApi implements IndicatorApiInterface {
-    const { code, builder, config } = indicator;
-    const buildAnalyticValues = this.getBuilderFunction(builder);
-    const analyticValues = await buildAnalyticValues({ api: this, config, fetchOptions });
+  /**
+   * An indicator may contain references to
+   * a. Other nested indicators
+   * b. "Primitive" elements (eg `dhis`, `tupaia` elements)
+   *
+   * Here we use the first category to compile a list of codes belonging to the second category,
+   * one nesting level at a time. We also include all nested indicators and all aggregations that
+   * will be used. This information is useful to other clients that fetch and build analytics
+   */
+  private getFetchAnalyticsDependencies = async (rootIndicatorCodes: string[]) => {
+    const buildersByNestDepth: Builder[][] = [];
+    const dataElementCodes = [];
+    const aggregations = [];
 
-    return analyticValues.map(value => ({ ...value, dataElement: code }));
-  };
+    let i;
+    let currentIndicatorCodes = rootIndicatorCodes;
+    for (i = 0; i < MAX_INDICATOR_NESTING_DEPTH; i++) {
+      const isRoot = i === 0;
+      const currentBuilders = await this.indicatorCodesToBuilders(currentIndicatorCodes, isRoot);
+      buildersByNestDepth.push(currentBuilders);
+      if (currentBuilders.length === 0) {
+        // No more indicators (root or nested)
+        break;
+      }
 
-  private getBuilderFunction = (builderName: string): Builder => {
-    const builderFunctionName = `build${upperFirst(builderName)}`;
-    if (!(builderFunctionName in builders)) {
-      throw new Error(`'${builderName}' is not an indicator builder`);
+      const { indicatorCodes, nonIndicatorCodes } = await this.getElementCodesByCategory(
+        currentBuilders,
+      );
+      currentIndicatorCodes = indicatorCodes;
+      dataElementCodes.push(...nonIndicatorCodes);
+      const currentAggregations = getAggregationsForBuilders(currentBuilders);
+      aggregations.push(...currentAggregations);
     }
 
-    return builders[builderFunctionName as keyof typeof builders];
+    if (i === MAX_INDICATOR_NESTING_DEPTH) {
+      // Avoid getting stuck in self-referencing indicators and cyclical references
+      throw new Error(`Max indicator nesting depth reached: ${MAX_INDICATOR_NESTING_DEPTH}`);
+    }
+
+    return {
+      buildersByNestDepth,
+      dataElementCodes: getUniqueEntries(dataElementCodes),
+      aggregations: getUniqueObjects(aggregations),
+    };
+  };
+
+  private indicatorCodesToBuilders = async (codes: string[], isRoot: boolean) => {
+    const indicators = await this.models.indicator.find({ code: codes });
+    return indicators.map(i => createBuilder(i, isRoot));
+  };
+
+  private getElementCodesByCategory = async (builders: Builder[]) => {
+    const indicatorCodes: string[] = [];
+    const nonIndicatorCodes: string[] = [];
+
+    const dataElements = await this.models.dataSource.findOrDefault({
```

Comment:
This is a bit dodgy here: we depend on `DataSource.findOrDefault`, which is specifically defined in the `DataSource` model in `@tupaia/data-broker`, and not in the base model in `@tupaia/database`.

It works in practice since `IndicatorApi` is currently instantiated in `DataBroker`, which passes in its own models. https://github.com/beyondessential/tupaia-backlog/issues/663 should solve this issue. Until then, we could approach this in various ways but I thought that leaving as is will be most probably fine in practice.